GWFish/modules/fishermatrix.py

```diff
@@ -19,80 +22,148 @@ def invertSVD(matrix):
-
-    # print(fisher_parameters[p] + ': ' + str(derivative))
-    return derivative
-
-
-def FisherMatrix(waveform, parameter_values, fisher_parameters, detector):
-
-    nd = len(fisher_parameters)
-    fm = np.zeros((nd, nd))
-
-    for p1 in np.arange(nd):
-        deriv1_p = fisher_parameters[p1]
-        deriv1 = derivative(waveform, parameter_values, deriv1_p, detector)
-        # sum Fisher matrices from different components of same detector (e.g., in the case of ET)
-        fm[p1, p1] = np.sum(aux.scalar_product(deriv1, deriv1, detector), axis=0)
-        for p2 in np.arange(p1+1, nd):
-            deriv2_p = fisher_parameters[p2]
-            deriv2 = derivative(waveform, parameter_values, deriv2_p, detector)
-            fm[p1, p2] = np.sum(aux.scalar_product(deriv1, deriv2, detector), axis=0)
-            fm[p2, p1] = fm[p1, p2]
-
-    return fm
-
+            pv = self.local_params[target_parameter]
+
+            dp = np.maximum(self.eps, self.eps * pv)
+
+            self.pv_set1 = self.local_params.copy()
+            self.pv_set2 = self.local_params.copy()
+            self.pv_set1[target_parameter] = pv - dp / 2.
+            self.pv_set2[target_parameter] = pv + dp / 2.
+
+            if target_parameter in ['ra', 'dec', 'psi']:  # these parameters do not influence the waveform
+    
+                signal1 = det.projection(self.pv_set1, self.detector, 
+                                         self.waveform_at_parameters[0], 
+                                         self.waveform_at_parameters[1])
+                signal2 = det.projection(self.pv_set2, self.detector, 
+                                         self.waveform_at_parameters[0], 
+                                         self.waveform_at_parameters[1])
+    
+                derivative = (signal2 - signal1) / dp
+            else:
+                # to improve precision of numerical differentiation
+                self.pv_set1['geocent_time'] = 0.
+                self.pv_set2['geocent_time'] = 0.
+                wave1, t_of_f1 = self.hphc_caller(self.waveform, self.pv_set1, 
+                                                  self.detector.frequencyvector)
+                wave2, t_of_f2 = self.hphc_caller(self.waveform, self.pv_set2, 
+                                                  self.detector.frequencyvector)
+                self.pv_set1['geocent_time'] = self.tc
+                self.pv_set2['geocent_time'] = self.tc
+                signal1 = det.projection(self.pv_set1, self.detector, wave1, t_of_f1 + self.tc)
+                signal2 = det.projection(self.pv_set2, self.detector, wave2, t_of_f2 + self.tc)
+    
+
+                derivative = np.exp(2j * np.pi * self.detector.frequencyvector \
+                                    * self.tc) * (signal2 - signal1) / dp
+
+        return derivative
+
+    def __call__(self, target_parameter):
+        return self.at_parameter(target_parameter)
+
+class FisherMatrix(object):
```

Comment:
Indeed, I think it would be useful to keep it as a class, to more easily add other methods in the future. For example, calculating the inverse. Some parts from `analyzeFisherErrors` can potentially be moved there too.